infra/lambda/datasets-endpoint/test_function.py

```diff
@@ -0,0 +1,307 @@
+"""
+Basic Lambda function tests. Working Data Lake AWS environment is required
+(run '$ cdk deploy' before running tests).
+"""
+
+import logging
+import re
+import uuid
+
+import function  # pylint:disable=import-error
+import pytest
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def test_request_no_method():
+    """Try request not containing method."""
+
+    body = {}
+
+    resp = function.lambda_handler({"body": body}, "context")
+    logger.info(resp)
+
+    assert resp["statusCode"] == 400
+    assert resp["body"]["message"] == "Bad Request: 'httpMethod' is a required property."
+
+
+def test_request_no_body():
+    """Try request not containing body."""
+
+    method = "POST"
+
+    resp = function.lambda_handler({"httpMethod": method}, "context")
+    logger.info(resp)
+
+    assert resp["statusCode"] == 400
+    assert resp["body"]["message"] == "Bad Request: 'body' is a required property."
+
+
+def test_post_method(db_truncate):  # pylint:disable=unused-argument
+    """Test Dataset creation using POST method."""
+
+    method = "POST"
+    body = {}
+    body["type"] = "RASTER"
+    body["title"] = "Dataset {}".format(uuid.uuid4().hex[:8])
+    body["owning_group"] = "A_XYZ_XYZ"
+
+    resp = function.lambda_handler({"httpMethod": method, "body": body}, "context")
+    logger.info(resp)
+
+    assert resp["statusCode"] == 201
+    assert resp["body"]["type"] == body["type"]
```

Comment:
One big issue with multiple assertions in a test is that it's not possible for more than one assertion to fail. For example, if the body and title are *both* wrong, only the first of them in the test will be reported. That makes the fix turnaround time longer, since we have to rerun the test after fixing one issue to (possibly) find that the other issue is still there. This can easily mask bigger issues under innocuous-looking test failures.

If test setup is really expensive, as in this case, asserting the full response can be an option, like `assert resp = {…}`. In a decent test framework like `unittest` something like `assertDictEqual(resp, {…})` will show a useful diff of the contents of the dictionaries on failure. Another alternative is to use [subtests](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest) to collect several independent test results in the same test function.

Reply:
I understand what you mean, but I think as it is now it more readable. When testing full response there will be lot of not relevant information.

Reply:
I suggest to leave it as is for now. What do you think @l0b0 ?

Reply:
Have you tried using subtests? Then at least we'll know about all the failures in one go, which is critical to fixing the issue quickly.